lib/data_stores/mysql_advanced_data_store.py

```diff
@@ -26,23 +26,48 @@ class Error(data_store.Error):
   """Base class for all exceptions in this module."""
 # pylint: enable=nonstandard-exception
 
+class SafeQueue(Queue.Queue):
+  def __init__(self, maxsize=0):
```

Comment:
Yeah, I stared at this one for a while before deciding to just override the entire init.  This was a really awful bug to dig out and test because you have to put the system under heavy load with a bunch of extra debug logging so you can see when it hits the deadlock.  This can take anywhere between 15 minutes and a few hours.  This fix isn't pretty, but it works.